homeassistant/components/mqtt/mixins.py

```diff
@@ -0,0 +1,471 @@
+                CONF_PAYLOAD_NOT_AVAILABLE: config[CONF_PAYLOAD_NOT_AVAILABLE],
+            }
+
+        if CONF_AVAILABILITY in config:
+            for avail in config[CONF_AVAILABILITY]:
+                self._avail_topics[avail[CONF_TOPIC]] = {
+                    CONF_PAYLOAD_AVAILABLE: avail[CONF_PAYLOAD_AVAILABLE],
+                    CONF_PAYLOAD_NOT_AVAILABLE: avail[CONF_PAYLOAD_NOT_AVAILABLE],
+                }
+
+        self._avail_config = config
+
+    async def _availability_subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def availability_message_received(msg: Message) -> None:
+            """Handle a new received MQTT availability message."""
+            topic = msg.topic
+            if msg.payload == self._avail_topics[topic][CONF_PAYLOAD_AVAILABLE]:
+                self._available = True
+            elif msg.payload == self._avail_topics[topic][CONF_PAYLOAD_NOT_AVAILABLE]:
+                self._available = False
+
+            self.async_write_ha_state()
+
+        topics = {}
+        for topic in self._avail_topics:
+            topics[f"availability_{topic}"] = {
+                "topic": topic,
+                "msg_callback": availability_message_received,
+                "qos": self._avail_config[CONF_QOS],
+            }
+
+        self._availability_sub_state = await async_subscribe_topics(
+            self.hass,
+            self._availability_sub_state,
+            topics,
+        )
+
+    @callback
+    def async_mqtt_connect(self):
+        """Update state on connection/disconnection to MQTT broker."""
+        if not self.hass.is_stopping:
+            self.async_write_ha_state()
+
+    async def async_will_remove_from_hass(self):
+        """Unsubscribe when removed."""
+        self._availability_sub_state = await async_unsubscribe_topics(
+            self.hass, self._availability_sub_state
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return if the device is available."""
+        if not self.hass.data[DATA_MQTT].connected and not self.hass.is_stopping:
+            return False
+        return not self._avail_topics or self._available
+
+
+async def cleanup_device_registry(hass, device_id):
+    """Remove device registry entry if there are no remaining entities or triggers."""
+    # Local import to avoid circular dependencies
+    # pylint: disable=import-outside-toplevel
+    from . import device_trigger, tag
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    entity_registry = await hass.helpers.entity_registry.async_get_registry()
+    if (
+        device_id
+        and not hass.helpers.entity_registry.async_entries_for_device(
+            entity_registry, device_id, include_disabled_entities=True
+        )
+        and not await device_trigger.async_get_triggers(hass, device_id)
+        and not tag.async_has_tags(hass, device_id)
```

Comment:
no need to await?

Reply:
It's not a coroutine function, `async` here just means it's async friendly and won't block.